authentication/ruby/azure/microsoft_kiota_authentication/lib/microsoft_kiota_authentication/azure_access_token_provider.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+require 'concurrent'
+require 'microsoft_kiota_abstractions'
+require 'oauth2'
+require_relative 'extensions/oauth2_ext'
+require_relative 'contexts/client_credential_context'
+require_relative 'contexts/authorization_code_context'
+require_relative 'contexts/on_behalf_of_context'
+
+module MicrosoftKiotaAuthentication
+  # Access Token Provider class implementation
+  class AzureAccessTokenProvider
+    include Concurrent::Async
+    # This is the initializer for AzureAccessTokenProvider.
+    # :params
+    #   token_request_context: a instance of one of our token request context
+    #   allowed_hosts: an array of strings, where each string is an allowed host, default is empty
+    #   scopes: an array of strings, where each string is a scope, default is empty array 
+    #   auth_code: a string containting the auth code; default is nil, can be updated post-initialization
+    def initialize(token_request_context, allowed_hosts = [], scopes = [])
+      raise StandardError, 'Parameter token_request_context cannot be nil.' if token_request_context.nil?
+
+      if !@token_request_context.instance_of?(MicrosoftKiotaAuthentication::ClientCredentialContext) && 
```

Comment:
we should probably define an interface instead (in this package, not in abstractions) so if people want to implement other flows, they can. That interface needs:
- get token method
- initialize auth provider method
- initialize scopes method
- auth provider getter